srml/contract/src/gas.rs

```diff
@@ -48,17 +90,42 @@ impl<T: Trait> GasMeter<T> {
 			limit: gas_limit,
 			gas_left: gas_limit,
 			gas_price,
+			#[cfg(test)]
+			tokens: Vec::new(),
 		}
 	}
 
 	/// Account for used gas.
 	///
+	/// Amount is calculated by the given `token`. If `token::calculate_amount` returns
+	/// `None` then all available gas is consumed and `OutOfGas` is returned.
+	///
 	/// Returns `OutOfGas` if there is not enough gas or addition of the specified
 	/// amount of gas has lead to overflow. On success returns `Proceed`.
 	///
-	/// NOTE that `amount` is always consumed, i.e. if there is not enough gas
+	/// NOTE that amount is always consumed, i.e. if there is not enough gas
 	/// then the counter will be set to zero.
-	pub fn charge(&mut self, amount: T::Gas) -> GasMeterResult {
+	#[inline]
+	pub fn charge<Tok: Token<T>>(
+		&mut self,
+		metadata: &Tok::Metadata,
+		token: Tok,
+	) -> GasMeterResult {
+		#[cfg(test)]
+		{
+			// Unconditionally add the token to the storage.
+			let erased_tok = ErasedToken {
+				description: format!("{:?}", token),
+				token: Box::new(token),
+			};
+			self.tokens.push(erased_tok);
+		}
+
+		let amount = match token.calculate_amount(metadata) {
+			Some(amount_in_gas) => amount_in_gas,
+			None => self.gas_left, // Consume everything
```

Comment:
If calculation of the amount fails, we just consume all remaining gas? That doesn't seem right. According to the docs above this could fail for a number of reasons, not only because it is out of gas.

Reply:
By above do you mean this one:

```
Returns `None` if the amount can't be calculated e.g. because of overflow.
```

?

If yes, I think I meant overflow and other similar cases. But now I can't come up of any of such similar case. Do you think rephrasing of docs (merely deleting e.g. and expanding the doc) could help here?